modules/tensor_mechanics/src/materials/ADViscoplasticityStressUpdate.C

```diff
@@ -0,0 +1,290 @@
+
+    residual += _intermediate_porosity * (h + _power_factor / h) - 1.0 -
+                _power_factor * Utility::pow<2>(_intermediate_porosity);
+    const ADReal dresidual_dh = _intermediate_porosity * (1.0 - _power_factor / Utility::pow<2>(h));
+    _derivative += dresidual_dh * dh_dM * dM_dtrial_gauge;
+  }
+
+  if (_verbose)
+  {
+    Moose::out << "in computeResidual:\n"
+               << "  position: " << _q_point[_qp] << " hydro_stress: " << _hydro_stress
+               << " equiv_stress: " << equiv_stress << " trial_grage: " << trial_gauge
+               << " M: " << M << std::endl;
+    Moose::out << "  residual: " << residual << "  derivative: " << _derivative << std::endl;
+  }
+
+  return residual;
+}
+
+template <ComputeStage compute_stage>
+ADReal
+ADViscoplasticityStressUpdate<compute_stage>::computeH(const Real n,
+                                                       const ADReal & M,
+                                                       const bool derivative)
+{
+  const ADReal mod = std::pow(M * _pore_shape_factor, (n + 1.0) / n);
+
+  // Calculate derivative with respect to M
+  if (derivative)
+  {
+    const ADReal dmod_dM = (n + 1.0) / n * mod / M;
+    return dmod_dM * std::pow(1.0 + mod / n, n - 1.0);
+  }
+
+  return std::pow(1.0 + mod / n, n);
+}
+
+template <ComputeStage compute_stage>
+ADRankTwoTensor
+ADViscoplasticityStressUpdate<compute_stage>::computeDGaugeDSigma(
+    const ADReal & gauge_stress,
+    const ADReal & equiv_stress,
+    const ADRankTwoTensor & dev_stress,
+    const ADRankTwoTensor & stress)
+{
+  // Compute the derivative of the gauge stress with respect to the equilvalent and hydrostatic
+  // stress components
+  const ADReal M = std::abs(_hydro_stress) / gauge_stress;
+  const ADReal h = computeH(_power, M);
+
+  // Compute the derviative of the residual with respect to the hydrostatic stress
+  ADReal dresidual_dhydro_stress = 0.0;
+  if (_hydro_stress)
+  {
+    const ADReal dM_dhydro_stress = M / _hydro_stress;
+    if (_model == ViscoplasticityModel::GTN)
+    {
+      dresidual_dhydro_stress = 2.0 * _intermediate_porosity * std::sinh(_pore_shape_factor * M) *
+                                _pore_shape_factor * dM_dhydro_stress;
+    }
+    else
+    {
+      const ADReal dresidual_dh =
+          _intermediate_porosity * (1.0 - _power_factor / Utility::pow<2>(h));
+      const ADReal dh_dM = computeH(_power, M, true);
+      dresidual_dhydro_stress = dresidual_dh * dh_dM * dM_dhydro_stress;
+    }
+  }
+
+  // Compute the derivative of the residual with respect to the equilvalent stress
+  ADReal dresidual_dequiv_stress = 2.0 * equiv_stress / Utility::pow<2>(gauge_stress);
+  if (_pore_shape == PoreShapeModel::SPHERICAL)
+    dresidual_dequiv_stress *= 1.0 + _intermediate_porosity / 1.5;
+
+  // Compute the derivative of the equilvalent stress to the deviatoric stress
+  const ADRankTwoTensor dequiv_stress_dsigma = 1.5 * dev_stress / equiv_stress;
```

Comment:
Here and elsewhere: are the methods in RankTwoTensor (like dsecondInvariant) not useful?  They at least account for things being zero.

Reply:
Probably...